spark2/src/main/java/org/apache/iceberg/spark/source/Reader.java

```diff
@@ -227,11 +246,17 @@ public StructType readSchema() {
     Broadcast<Table> tableBroadcast = sparkContext.broadcast(SerializableTable.copyOf(table));
 
     List<InputPartition<InternalRow>> readTasks = Lists.newArrayList();
-    for (CombinedScanTask task : tasks()) {
-      readTasks.add(new ReadTask<>(
-          task, tableBroadcast, expectedSchemaString, caseSensitive,
-          localityPreferred, InternalRowReaderFactory.INSTANCE));
-    }
+    Tasks.foreach(tasks())
+        .stopOnFailure()
+        .executeWith(readTasksInitExecutorService)
+        .run(task -> {
+          InputPartition<InternalRow> readTask = new ReadTask<>(
+              task, tableBroadcast, expectedSchemaString, caseSensitive,
+              localityPreferred, InternalRowReaderFactory.INSTANCE);
+          synchronized (readTasks) {
+            readTasks.add(readTask);
```

Comment:
Maybe it‘s better to use array instead of list, and then we can avoid to add lock or temporary var ?

Reply:
Yes, using array can help avoid lock, but the lock time should be insignificant compared to the get block location operation. The return value is a list and I didn't want to do all the transformation.

Reply:
Yeap! But another reason is we can keep the style consistent with spark3 module, a little bit of optimization with array as follow  [:R153](https://github.com/apache/iceberg/pull/2800/files#diff-47b9fbed0d32a8f8ec128a4f78cd93f238e65c38f93e9a10d257eb83838563ffR153), how do you think?

Reply:
Okay!